google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkPreconditionsSupport.java

```diff
@@ -0,0 +1,199 @@
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.intellij.appengine.sdk;
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import com.google.cloud.tools.intellij.appengine.sdk.CloudSdkService.SdkStatus;
+import com.google.cloud.tools.intellij.flags.PropertiesFileFlagReader;
+import com.google.cloud.tools.intellij.util.GctBundle;
+import com.google.common.annotations.VisibleForTesting;
+import com.intellij.notification.Notification;
+import com.intellij.notification.NotificationType;
+import com.intellij.notification.Notifications;
+import com.intellij.openapi.actionSystem.AnAction;
+import com.intellij.openapi.actionSystem.AnActionEvent;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.options.ShowSettingsUtil;
+import com.intellij.openapi.progress.ProgressIndicator;
+import com.intellij.openapi.progress.ProgressManager;
+import com.intellij.openapi.progress.Task;
+import com.intellij.openapi.project.Project;
+import com.intellij.remoteServer.runtime.deployment.ServerRuntimeInstance.DeploymentOperationCallback;
+import com.intellij.remoteServer.runtime.log.LoggingHandler;
+import java.util.concurrent.CountDownLatch;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Utility class providing support for blocking / tracking {@link CloudSdkService#install()} process
+ * and other SDK preconditions so that dependent deployment processes can be postponed until SDK is
+ * completely ready.
+ */
+public class CloudSdkPreconditionsSupport {
+  private static CloudSdkPreconditionsSupport instance;
+
+  public static CloudSdkPreconditionsSupport getInstance() {
+    if (instance == null) {
+      instance = new CloudSdkPreconditionsSupport();
+    }
+
+    return instance;
+  }
+
+  /**
+   * Waits for Cloud SDK to be ready for all operations and then runs the given runnable. If process
+   * results in error or user cancel, shows notification and does not run.
+   *
+   * @param project Project to which runnable belongs.
+   * @param runnable Runnable to run after Cloud SDK is ready.
+   * @param deploymentLog Log to print SDK statuses.
+   * @param callback Callback to report errors.
+   */
+  public void runAfterCloudSdkPreconditionsMet(
+      @Nullable Project project,
+      Runnable runnable,
+      LoggingHandler deploymentLog,
+      DeploymentOperationCallback callback) {
+    CloudSdkService cloudSdkService = CloudSdkService.getInstance();
+
+    SdkStatus sdkStatus = cloudSdkService.getStatus();
+    boolean installInProgress = sdkStatus == SdkStatus.INSTALLING;
+    // if not already installing and still not ready, attempt to fix and install now.
+    if (!installInProgress && sdkStatus != SdkStatus.READY && cloudSdkService.supportsInstall()) {
+      cloudSdkService.install();
+      installInProgress = true;
+    }
+
+    CountDownLatch installationCompletionLatch = new CountDownLatch(1);
```

Comment:
It definitely could be done this way, but since we need to poll `ProgressManager` for cancellation status, we'll end up spinning a loop checking for this boolean value and then sleeping for a while, and `await` method is a bit cleaner to understand maybe. It almost feel the same way, but maybe `CountDownLatch` a bit better implies what exactly the deployment is waiting for by looking at the listener code.

Reply:
I see. sgtm.

so now you are blocking for 1 sec, then polling to see if the sdk installation is cancelled. For the user, this means that they (just to make sure I understand):

1) click cancel to stop the sdk installation
2) a max of 1 second later the user will be notified that the deployment failed

Is that correct?

Reply:
Yep, correct. Better way would be to trigger the latch on cancel itself, but I didn't initially find the way to properly do this, `onCancel()` method in `Task` does not get called.

Reply:
I may just be misunderstanding something with the last comment but, the `checkIfCanceled` in your latch count loop is checking if the cloud SDK installation is canceled, right? Not if deployment is canceled.

Reply:
It does check if deployment is canceled - it checks the progress instance assigned to this `Task` thread, where we are waiting for Cloud SDK to become ready.

Reply:
Oh. my bad, I think I had this confused. wdyt about a quick review meeting after you do any pending refactorings to discuss in more detail?